lib/phase/functional.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+const Joi = require('joi');
+const Hoek = require('hoek');
+const clone = require('clone');
+
+// Actual environment size is limited by space, not quantity.
+// We do this to force sd.yaml to be simpler.
+const MAX_ENVIRONMENT_VARS = 25;
+// If they are trying to execute >25 permutations,
+// maybe there is a better way to accomplish their goal.
+const MAX_PERMUTATIONS = 25;
+
+/**
+ * Functional Phase
+ *  - Validates Jobs have steps and not too many env. variables
+ *  - Validates Workflow starts with main and contains all the jobs
+ *  - Validates Matrix isn't too big
+ * @method
+ * @param  {Object}   flattenedDoc Document that went through flattening
+ * @param  {Function} callback     Function to call when done (err, doc)
+ */
+module.exports = (flattenedDoc, callback) => {
+    const doc = flattenedDoc;
+    const errors = [];
+    const listOfJobs = Object.keys(doc.jobs);
+
+    // Jobs
+    const SCHEMA_JOB = Joi.object()
+        .keys({
+            steps: Joi.object().min(1),
+            image: Joi.string().required(),
+            environment: Joi.object().max(MAX_ENVIRONMENT_VARS)
+        })
+        .unknown(true)
+        // Add documentation
+        .options({
+            language: {
+                object: {
+                    min: 'requires at least one step',
+                    max: `can only have ${MAX_ENVIRONMENT_VARS} environment variables defined`
+                }
+            }
+        });
+
+    // Validate jobs contain required minimum fields
```

Comment:
I suggest for modularity and readability sake, turn every one of these validation checks into a function that returns an array (of error messages) and concat them together that way